deltachat-ffi/deltachat.h

```diff
@@ -2630,6 +2636,106 @@ char* dc_get_last_error (dc_context_t* context);
 void dc_str_unref (char* str);
 
 
+/**
+ * @class dc_backup_provider_t
+ *
+ * Set up another device.
+ */
+
+/**
+ * Creates an object for sending a backup to another device.
+ *
+ * The backup is sent to through a peer-to-peer channel which is bootstrapped
+ * by a QR-code.  The backup contains the entire state of the account
+ * including credentials.  This can be used to setup a new device.
+ *
+ * Once this function returns the backup is being offered to remove devices.
+ * To wait until one device transferred the backup, use
+ * dc_backup_provider_done().  Alternatively abort the operation using
+ * dc_stop_ongoing_process().
+ *
+ * During execution of the job #DC_EVENT_IMEX_PROGRESS is sent out to indicate
+ * state and progress.
+ *
+ * @memberof dc_backup_sender_t
+ * @param context The context.
+ * @return Opaque object for sending the backup.
+ *    On errors, NULL is returned and dc_get_last_error()returns an error that
+ *    should be shown to the user.
+ */
+dc_backup_provider_t* dc_provide_backup (dc_context_t* context);
+
+
+/**
+ * Returns the QR code text that will offer the backup to other devices.
+ *
+ * The QR code contains a ticket which will validate the backup and provide
+ * authentication for both the provider and the recipient.
+ *
+ * The scanning device should call the scanned text to dc_check_qr().  If
+ * dc_check_qr() returns DC_QR_BACKUP the backup transfer can be started using
+ * dc_get_backup().
+ *
+ * @memberof dc_backup_provider_t
+ * @param context The context.
+ * @param backup_provider The backup provider object as created by
+ *    dc_provide_backup().
+ * @return The text that should be put in the QR code.
+ *    On errors and empty QR code is returned, NULL is never returned.
+ *    the returned string must be released using dc_str_unref() after usage.
+ */
+char* dc_backup_provider_qr (dc_context_t* context, const dc_backup_provider_t* backup_provider);
+
+
+/**
+ * Returns the QR code SVG image that will offer the backup to other devices.
+ *
+ * This works like dc_backup_provider_qr() but returns the text of a rendered
+ * SVG image containing the QR code.
+ *
+ * @memberof dc_backup_provider_t
+ * @param context The context.
+ * @param backup_provider The backup provider object as created by
+ *    dc_provide_backup().
+ * @return The text that should be put in the QR code.
+ *    On errors and empty QR code is returned, NULL is never returned.
+ *    the returned string must be released using dc_str_unref() after usage.
+ */
+char * dc_backup_provider_qr_svg (dc_context_t* context, const dc_backup_provider_t* backup_provider);
+
+/**
+ * Waits for the sending to finish and frees the backup provider object.
+ *
+ * @memberof dc_backup_sender_t
+ * @param context The context.
+ * @param backup_sender The backup sender object as created by dc_send_backup().
+ *    If NULL is given nothing is done.
+ */
+void dc_backup_provider_done (dc_context_t* context, dc_backup_provider_t* backup_provider);
```

Comment:
I'm a bit torn on `dc_backup_provider_unref()` but decided to implement this anyway, so now there's a `_wait` and `_unref`.

Now `BackupProvider` implements `Future` directly so it can be awaited without having to move it, this is similar to how tokio's `JoinHandle` works.  I considered making it a `FusedFuture` so that the FFI could be made a little safer by checking that the future is still running.  I ended up not doing that because the `deltachat-ffi` crate doesn't yet have a dependency on `futures` and it felt a bit insignificant to add a new dependency.